File: android/src/app/src/main/java/com/example/agitodo/OAuth.java
package com.example.agitodo;

import java.util.List;
import java.util.ArrayList;

import android.net.Uri;

import org.apache.http.NameValuePair;
import org.apache.http.message.BasicNameValuePair;

public class OAuth {
    static final String OAUTH_REDIRECT_URI = "http://localhost:8081/oauth-code";

    static final String DROPBOX_APP_KEY = "DROPBOX_APP_KEY";
    static final String DROPBOX_APP_SECRET = "DROPBOX_APP_SECRET";

    static final String GOOGLE_CLIENT_ID = "GOOGLE_CLIENT_ID";
    static final String GOOGLE_CLIENT_SECRET = "GOOGLE_CLIENT_SECRET";

    static final String HUBIC_CLIENT_ID = "HUBIC_CLIENT_ID";
    static final String HUBIC_CLIENT_SECRET = "HUBIC_CLIENT_SECRET";

    static final String DROPBOX_AUTH_ENDPOINT = "https://www.dropbox.com/1/oauth2/authorize";
    static final String GOOGLE_AUTH_ENDPOINT = "https://accounts.google.com/o/oauth2/auth";
    static final String HUBIC_AUTH_ENDPOINT = "https://api.hubic.com/oauth/auth";

    static final String DROPBOX_TOKEN_ENDPOINT = "https://api.dropbox.com/1/oauth2/token";
    static final String GOOGLE_TOKEN_ENDPOINT = "https://accounts.google.com/o/oauth2/token";
    static final String HUBIC_TOKEN_ENDPOINT = "https://api.hubic.com/oauth/token";

    static final int PROVIDER_DROPBOX = 0;
    static final int PROVIDER_GDRIVE = 1;
    static final int PROVIDER_GMAIL = 2;
    static final int PROVIDER_HUBIC = 3;

    static Logger log = new Logger();

    int provider;
    String client_id;
    String client_secret;

    public OAuth(int provider) {
        this.provider = provider;

        switch (provider) {
            case PROVIDER_DROPBOX:
                this.client_id = DROPBOX_APP_KEY;
                this.client_secret = DROPBOX_APP_SECRET;
                break;
            case PROVIDER_GDRIVE:
            case PROVIDER_GMAIL:
                this.client_id = GOOGLE_CLIENT_ID;
                this.client_secret = GOOGLE_CLIENT_SECRET;
                break;
            case PROVIDER_HUBIC:
                this.client_id = HUBIC_CLIENT_ID;
                this.client_secret = HUBIC_CLIENT_SECRET;
                break;
        }
    }

    public String authURL() {
        String endpoint = "";
        List<NameValuePair> params = new ArrayList<NameValuePair>();

        params.add(new BasicNameValuePair("response_type", "code"));
        params.add(new BasicNameValuePair("redirect_uri", OAUTH_REDIRECT_URI));
        params.add(new BasicNameValuePair("client_id", client_id));

        switch (provider) {
            case PROVIDER_DROPBOX:
                endpoint = DROPBOX_AUTH_ENDPOINT;
                params.add(new BasicNameValuePair("force_reapprove", "true"));
                break;
            case PROVIDER_GDRIVE:
                endpoint = GOOGLE_AUTH_ENDPOINT;
                params.add(new BasicNameValuePair("scope",
                        "https://www.googleapis.com/auth/drive.appdata"));
                params.add(new BasicNameValuePair("include_granted_scopes", "false"));
                break;
            case PROVIDER_GMAIL:
                endpoint = GOOGLE_AUTH_ENDPOINT;
                params.add(new BasicNameValuePair("scope",
                        "https://www.googleapis.com/auth/gmail.compose"));
                params.add(new BasicNameValuePair("include_granted_scopes", "false"));
                break;
            case PROVIDER_HUBIC:
                endpoint = HUBIC_AUTH_ENDPOINT;
                params.add(new BasicNameValuePair("scope", "credentials.r"));
                break;
        }

        return endpoint + "?" + Request.encodeQuery(params);
    }

    public Request.Response token(String code, boolean refresh) {
        String endpoint = "";
        List<NameValuePair> params = new ArrayList<NameValuePair>();

        params.add(new BasicNameValuePair("client_id", client_id));
        params.add(new BasicNameValuePair("client_secret", client_secret));

        if (refresh) {
            params.add(new BasicNameValuePair("refresh_token", code));
            params.add(new BasicNameValuePair("grant_type", "refresh_token"));
        } else {
            params.add(new BasicNameValuePair("code", code));
            params.add(new BasicNameValuePair("grant_type", "authorization_code"));
            params.add(new BasicNameValuePair("redirect_uri", OAUTH_REDIRECT_URI));
        }

        switch (provider) {
            case PROVIDER_DROPBOX:
                endpoint = DROPBOX_TOKEN_ENDPOINT;
                break;
            case PROVIDER_GDRIVE:
            case PROVIDER_GMAIL:
                endpoint = GOOGLE_TOKEN_ENDPOINT;
                break;
            case PROVIDER_HUBIC:
                endpoint = HUBIC_TOKEN_ENDPOINT;
                break;
        }

        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("Content-Type",
                "application/x-www-form-urlencoded"));

        Uri uri = Uri.parse(endpoint);

        return new Request(Request.POST, uri.getHost(), uri.getPath(), headers,
                null).send(Request.encodeQuery(params));
    }
}

File: www/oauth.js
var crypto = require("crypto");
var querystring = require("querystring");
var url = require("url");

var request = require("./request");

var ALIAS_EXPIRES = 45 * 60000; // min -> ms
var ALIAS_LENGTH = 10; // bytes

var DROPBOX_APP_KEY = "DROPBOX_APP_KEY";
var DROPBOX_APP_SECRET = "DROPBOX_APP_SECRET";

var GOOGLE_CLIENT_ID = "GOOGLE_CLIENT_ID";
var GOOGLE_CLIENT_SECRET = "GOOGLE_CLIENT_SECRET";
var GOOGLE_CLIENT_ID_DEBUG = "GOOGLE_CLIENT_ID_DEBUG";
var GOOGLE_CLIENT_SECRET_DEBUG = "GOOGLE_CLIENT_SECRET_DEBUG";

var HUBIC_CLIENT_ID = "HUBIC_CLIENT_ID";
var HUBIC_CLIENT_SECRET = "HUBIC_CLIENT_SECRET";
var HUBIC_CLIENT_ID_DEBUG = "HUBIC_CLIENT_ID_DEBUG";
var HUBIC_CLIENT_SECRET_DEBUG = "HUBIC_CLIENT_SECRET_DEBUG";

var OAUTH_REDIRECT_URI = "https://example.com/oauth-code";
var OAUTH_REDIRECT_URI_DEBUG = "http://localhost:8081/oauth-code";

var DROPBOX_AUTH_ENDPOINT = "https://www.dropbox.com/1/oauth2/authorize";
var GOOGLE_AUTH_ENDPOINT = "https://accounts.google.com/o/oauth2/auth";
var HUBIC_AUTH_ENDPOINT = "https://api.hubic.com/oauth/auth";

var DROPBOX_TOKEN_ENDPOINT = "https://api.dropbox.com/1/oauth2/token";
var GOOGLE_TOKEN_ENDPOINT = "https://accounts.google.com/o/oauth2/token";
var HUBIC_TOKEN_ENDPOINT = "https://api.hubic.com/oauth/token";

var SERVICE = {
  DROPBOX: 0,
  GDRIVE: 1,
  GMAIL: 2,
  HUBIC: 3,
  0: "DROPBOX",
  1: "GDRIVE",
  2: "GMAIL",
  3: "HUBIC"
};

// Prototype: Dropbox
function Dropbox(data) {
  var data = (Object.prototype.toString.call(data) === "[object Object]") ?
    data : {};

  this.access_token = (typeof data.access_token === "string") ? data.access_token :
    "";
  this.access_token_expires = (typeof data.access_token_expires === "number") ?
    data.access_token_expires : -1;
};

// Prototype: Google
function Google(data) {
  var data = (Object.prototype.toString.call(data) === "[object Object]") ?
    data : {};

  this.access_token = (typeof data.access_token === "string") ? data.access_token :
    "";
  this.access_token_expires = (typeof data.access_token_expires === "number") ?
    data.access_token_expires : -1;
  this.refresh_token = (typeof data.refresh_token === "string") ? data.refresh_token :
    "";
};

// Prototype: Hubic
function Hubic(data) {
  var data = (Object.prototype.toString.call(data) === "[object Object]") ?
    data : {};

  this.access_token = (typeof data.access_token === "string") ? data.access_token :
    "";
  this.access_token_expires = (typeof data.access_token_expires === "number") ?
    data.access_token_expires : -1;
  this.refresh_token = (typeof data.refresh_token === "string") ? data.refresh_token :
    "";
  this.openstack_token = (typeof data.openstack_token === "string") ? data.openstack_token :
    "";
  this.openstack_token_expires = (typeof data.openstack_token_expires ===
      "number") ?
    data.openstack_token_expires : -1;
  this.openstack_endpoint = (typeof data.openstack_endpoint === "string") ?
    data.openstack_endpoint :
    "";
};

var alias = (function() {
  var cache = {};

  function removeExpired() {
    var now = new Date().getTime();

    for (var i in cache) {
      if ((now - cache[i].time) >= ALIAS_EXPIRES) {
        delete cache[i];
      }
    }
  }

  function pop(token) {
    if (typeof token !== "string" || token.length > (ALIAS_LENGTH * 2) ||
      !cache.hasOwnProperty(token)) {
      return;
    }

    var item = cache[token];
    delete cache[token];

    return ((new Date().getTime() - item.time) < ALIAS_EXPIRES) ? item.sid :
      undefined;
  }

  function push(sid) {
    removeExpired();

    var token;

    do {
      token = crypto.randomBytes(ALIAS_LENGTH).toString("hex");
    } while (token in cache);

    cache[token] = {
      sid: sid,
      time: new Date().getTime()
    };

    return token;
  }

  return {
    push: push,
    pop: pop
  };
})();

function authURL(session, service) {
  var endpoint;
  var params = {
    response_type: "code",
    redirect_uri: session.debug ? OAUTH_REDIRECT_URI_DEBUG : OAUTH_REDIRECT_URI
  };

  switch (service) {
    case SERVICE.DROPBOX:
      endpoint = DROPBOX_AUTH_ENDPOINT;
      params.client_id = DROPBOX_APP_KEY;
      params.force_reapprove = "true";
      break;
    case SERVICE.GDRIVE:
      endpoint = GOOGLE_AUTH_ENDPOINT;
      params.client_id = session.debug ? GOOGLE_CLIENT_ID_DEBUG :
        GOOGLE_CLIENT_ID;
      params.scope = "https://www.googleapis.com/auth/drive.appdata";
      params.include_granted_scopes = "false";
      if (!session.debug) {
        params.access_type = "offline";
        params.approval_prompt = "force";
      }
      break;
    case SERVICE.GMAIL:
      endpoint = GOOGLE_AUTH_ENDPOINT;
      params.client_id = session.debug ? GOOGLE_CLIENT_ID_DEBUG :
        GOOGLE_CLIENT_ID;
      params.scope = "https://www.googleapis.com/auth/gmail.compose";
      params.include_granted_scopes = "false";
      if (!session.debug) {
        params.access_type = "offline";
        params.approval_prompt = "force";
      }
      break;
    case SERVICE.HUBIC:
      endpoint = HUBIC_AUTH_ENDPOINT;
      params.client_id = session.debug ? HUBIC_CLIENT_ID_DEBUG :
        HUBIC_CLIENT_ID;
      params.scope = "credentials.r";
      break;
    default:
      return "/";
  }

  params.state = alias.push(session.id);
  session.oauth.state = params.state;

  return endpoint + "?" + querystring.stringify(params);
}

function token(session, service, code, refresh, callback) {
  if (typeof code !== "string") {
    return callback ? callback() : undefined;
  }

  var endpoint;
  var params = {};

  if (refresh) {
    params.refresh_token = code;
    params.grant_type = "refresh_token";
  } else {
    params.code = code;
    params.grant_type = "authorization_code";
    params.redirect_uri = session.debug ? OAUTH_REDIRECT_URI_DEBUG :
      OAUTH_REDIRECT_URI;
  }

  switch (service) {
    case SERVICE.DROPBOX:
      endpoint = DROPBOX_TOKEN_ENDPOINT;
      params.client_id = DROPBOX_APP_KEY;
      params.client_secret = DROPBOX_APP_SECRET;
      break;
    case SERVICE.GDRIVE:
    case SERVICE.GMAIL:
      endpoint = GOOGLE_TOKEN_ENDPOINT;
      params.client_id = session.debug ? GOOGLE_CLIENT_ID_DEBUG :
        GOOGLE_CLIENT_ID;
      params.client_secret = session.debug ? GOOGLE_CLIENT_SECRET_DEBUG :
        GOOGLE_CLIENT_SECRET;
      break;
    case SERVICE.HUBIC:
      endpoint = HUBIC_TOKEN_ENDPOINT;
      params.client_id = session.debug ? HUBIC_CLIENT_ID_DEBUG :
        HUBIC_CLIENT_ID;
      params.client_secret = session.debug ? HUBIC_CLIENT_SECRET_DEBUG :
        HUBIC_CLIENT_SECRET;
      break;
    default:
      return callback ? callback() : undefined;
  }

  var body = querystring.stringify(params);

  var headers = {
    "Content-Type": "application/x-www-form-urlencoded"
  };

  var u = url.parse(endpoint);

  request("POST", u.hostname, u.pathname, headers, null, body, callback);
}

module.exports = {
  alias: alias,
  authURL: authURL,
  token: token,
  SERVICE: SERVICE,
  Dropbox: Dropbox,
  Google: Google,
  Hubic: Hubic
};
